File: apps/bff/src/platform/ordering.ts
import { generateKeyBetween } from 'fractional-indexing';

/**
 * 수동 정렬 키 — Figma 방식 fractional indexing.
 *
 * float 중간값 방식과 달리 키가 문자열이라 자릿수를 필요한 만큼 늘릴 수 있어
 * 어떤 순서로 몇 번을 끼워 넣어도 "사이값이 안 만들어지는" 정밀도 고갈이 없다.
 * 키는 ASCII 알파벳으로만 구성되며 바이트 순서 비교를 전제하므로,
 * DB 컬럼은 반드시 `text collate "C"`여야 한다 (로케일 collation 금지).
 */

/** 리스트 맨 뒤에 붙일 키. last가 null이면 첫 키. */
export function orderKeyAfter(last: string | null): string {
  return generateKeyBetween(last, null);
}

/** a와 b 사이에 끼울 키. a=null은 맨 앞, b=null은 맨 뒤. */
export function orderKeyBetween(a: string | null, b: string | null): string {
  return generateKeyBetween(a, b);
}

File: experiments/bff-rs/src/ordering.rs
//! 수동 정렬 키. `apps/bff/src/platform/ordering.ts`(fractional-indexing) 대응.
//!
//! 원본은 fractional-indexing으로 임의의 두 키 "사이"에 끼울 수 있지만, 인메모리
//! 스토어는 리스트 맨 뒤 append(`orderKeyAfter(lastPosition)`)만 쓴다. 여기선 그
//! append 경로만 충실히 옮긴다 — 'a'..='z' 위에서 마지막 글자를 올리거나('a'->'b'),
//! 'z'면 한 글자 늘려('z'->'za') 항상 사전순으로 엄격히 증가하는 키를 만든다.

/// 리스트 맨 뒤에 붙일 키. `last`가 없으면 첫 키("a").
pub fn key_after(last: Option<&str>) -> String {
    match last {
        None => "a".to_string(),
        Some(s) => {
            let mut bytes = s.as_bytes().to_vec();
            match bytes.last_mut() {
                Some(c) if *c < b'z' => {
                    *c += 1;
                    String::from_utf8(bytes).expect("ascii")
                }
                _ => {
                    let mut out = s.to_string();
                    out.push('a');
                    out
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strictly_increasing_appends() {
        let mut last: Option<String> = None;
        let mut prev = String::new();
        for _ in 0..100 {
            let k = key_after(last.as_deref());
            assert!(k > prev, "expected {k:?} > {prev:?}");
            prev = k.clone();
            last = Some(k);
        }
    }

    #[test]
    fn rolls_over_at_z() {
        assert_eq!(key_after(None), "a");
        assert_eq!(key_after(Some("a")), "b");
        assert_eq!(key_after(Some("y")), "z");
        assert_eq!(key_after(Some("z")), "za");
        assert_eq!(key_after(Some("za")), "zb");
    }
}
